src/runtime/pkg/kata-monitor/cri.go

```diff
@@ -0,0 +1,202 @@
+	if err != nil {
+		errMsg := errors.Wrapf(err, "connect endpoint '%s', make sure you are running as root and the endpoint has been started", endPoint)
+		return nil, errMsg
+	}
+	monitorLog.Debugf("connected successfully using endpoint: %s", endPoint)
+	return conn, nil
+}
+
+func matchesRegex(pattern, target string) bool {
+	if pattern == "" {
+		return true
+	}
+	matched, err := regexp.MatchString(pattern, target)
+	if err != nil {
+		// Assume it's not a match if an error occurs.
+		return false
+	}
+	return matched
+}
+
+func closeConnection(conn *grpc.ClientConn) error {
+	if conn == nil {
+		return nil
+	}
+	return conn.Close()
+}
+
+func getRuntimeClient(runtimeEndpoint string) (pb.RuntimeServiceClient, *grpc.ClientConn, error) {
+	var (
+		conn *grpc.ClientConn
+		err  error
+	)
+	// Set up a connection to the server.
+	// If no EndPoint set then use the default endpoint types
+	conn, err = getConnection(runtimeEndpoint)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	runtimeClient := pb.NewRuntimeServiceClient(conn)
+	return runtimeClient, conn, nil
+}
+
+func dial(ctx context.Context, addr string) (net.Conn, error) {
+	return (&net.Dialer{}).DialContext(ctx, unixProtocol, addr)
+}
+
+func parseEndpointWithFallbackProtocol(endpoint string, fallbackProtocol string) (protocol string, addr string, err error) {
+	if protocol, addr, err = parseEndpoint(endpoint); err != nil && protocol == "" {
+		fallbackEndpoint := fallbackProtocol + "://" + endpoint
+		protocol, addr, err = parseEndpoint(fallbackEndpoint)
+		if err == nil {
+			monitorLog.Warningf("Using %q as endpoint is deprecated, please consider using full url format %q.", endpoint, fallbackEndpoint)
+		}
+	}
+	return
+}
+
+func parseEndpoint(endpoint string) (string, string, error) {
+	u, err := url.Parse(endpoint)
+	if err != nil {
+		return "", "", err
+	}
+
+	switch u.Scheme {
+	case "tcp":
+		return "tcp", u.Host, nil
+
+	case "unix":
+		return "unix", u.Path, nil
+
+	case "":
+		return "", "", fmt.Errorf("using %q as endpoint is deprecated, please consider using full url format", endpoint)
+
+	default:
+		return u.Scheme, "", fmt.Errorf("protocol %q not supported", u.Scheme)
+	}
+}
+
+// getSandboxes get kata sandbox from the container engine.
+// this will be called only after monitor start.
+func (km *KataMonitor) getSandboxes() (map[string]struct{}, error) {
+
+	sandboxMap := make(map[string]struct{})
+	runtimeClient, runtimeConn, err := getRuntimeClient(km.runtimeEndpoint)
+	if err != nil {
+		return sandboxMap, err
+	}
+	defer closeConnection(runtimeConn)
+
+	filter := &pb.PodSandboxFilter{
+		State: &pb.PodSandboxStateValue{
+			State: pb.PodSandboxState_SANDBOX_READY,
+		},
+	}
+
+	request := &pb.ListPodSandboxRequest{
+		Filter: filter,
+	}
+	monitorLog.Debugf("ListPodSandboxRequest: %v", request)
+	r, err := runtimeClient.ListPodSandbox(context.Background(), request)
+	monitorLog.Debugf("ListPodSandboxResponse: %v", r)
```

Comment:
Do we need to also log the error here? Or should this line be moved to after the check on `err`?

Reply:
Yeah, in case of error r will likely be nil. Moved after the check on err.